### PR TITLE
Generate: return `past_key_values`

### LINKED ISSUE #24841: Support for caching prompt hidden states through multiple calls of `generate()`
### Feature request

Hi there,

I'd like to be able to re-use the hidden states for a common (potentially long) prompt across multiple calls to `model.generate()` in order to reduce redundant computation. Here is how I envision a final API, though I'm sure there are multiple ways to do it.

```python
# Load stuff
model = AutoModel.from_pretrained('huggyllama/llama-7b')
tokenizer = AutoTokenizer.from_pretrained('huggyllama/llama-7b')

# Common prompt that we'd prepend to every example
prompt = "This is a common prompt in every example."
prompt_ids = tokenizer(prompt, return_tensors='pt')

# Examples to pass to generate
examples = ["Ackbar went to", "Billaba enjoys", "Cody is eating some"]

# Generation loop
outputs = []
prompt_hidden_state = None
for ex in examples:
    # Current way of doing things
    out = model.generate(
        **tokenizer(prompt + ex, return_tensors='pt'),   
    )

    # Proposed method to re-use prompt_hidden_state
    out = model.generate(
        **tokenizer(x, return_tensors='pt'),   
        common_prompt_ids=prompt_ids,
        prompt_hidden_state=prompt_hidden_state
    )
    prompt_hidden_state = out.prompt_hidden_state
    outputs.append(out.sequences)
```

Thanks in advance. 

### Motivation

A very common pattern for LLM usage is having a common prompt (e.g., instructions and input/output pairs), a sample input, and asking it to generate the sample output. For example:

```
You are a programmer's assistant which converts English descriptions to Python functions.

English: <example 1 description>
Python: <example 1 function>

English: <example 2 description>
Python: <example 2 function>

English: <example 3 description>
Python: <example 3 function>

English: <input description>
Python: 
```

I'd like to be able to cache the common part of the prompt across inputs, that is, everything before `<input description>` which appears in every example to avoid potentially expensive re-computation.

### Your contribution

The only existing info I could find is the short discussion [here](https://discuss.huggingface.co/t/avoid-recalculating-hidden-states-between-generate-calls/34209). I tried messing around a bit to get this to work but had little luck. I'm not familiar with the inner-workings of `transformers` and ran into numerous errors. One problem is padding, which if we're using left padding, can cause some misalignment with the prompt hidden states, e.g.:
```
<p> <p> <p> common prompt x_1 x_2 x_3
<p> <p> common prompt x_1 x_2 x_3 x_4
<p> <p> <p> <p> common prompt x_1 x_2
```
I don't know the best way to solve this. Do we dynamically pad every tensor in `past_key_values`? That seems slow but I don't know if it actually is.

If someone can suggest a better/easier way or maybe give some more pointers on how to solve padding. I'd be happy to try again myself.

Thanks in advance.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Enables returning `past_key_values` from `generate`, if `return_dict_in_generate=True` (otherwise only the generated `input_ids` are returned) and `use_cache=True` (otherwise there is no cache to return ;) ).

In more abstract terms, this enables features like:
1. continuing a given generation without having the more expensive prefill step -- like in multi-turn conversations
2. exploring the KV values without having to place a breakpoint in `generate` 👀 🐛 

The added code for the feature is minimal, so most of the PR is docs and tests 🤗  

Fixes #24841

### CODE REVIEW

**Comment by freckletonj** [comment]
This is a killer feature :+1:

**Comment by kazzand** [comment]
@gante Hi! Thanks for PR!
Did you test feeding output past_key_values into .generate() method? Like take first 250 tokens input, run .generate(), get output past_key_values, take another 50 tokens input and run .generate() with previous 250 past_key_values? With beam search it seems to be kinda tricky. I'm trying to resolve multiple dimension mismatch problems.

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1822,6 +1822,85 @@ def test_generate_from_inputs_embeds_decoder_only(self):
...
+
+            model = model_class(config).to(torch_device)
+            model.eval()
+            model.generation_config.pad_token_id = model.generation_config.eos_token_id = -1
+            model.generation_config.forced_eos_token_id = None
+
+            # If "past_key_values" is not returned, pass the test (e.g. RWKV uses a different cache name and format)
+            outputs = model(**inputs)
+            if "past_key_values" not in outputs:
+                return
```
Comment (amyeroberts) [inline tests/generation/test_utils.py:1861]: I think that rather than have model specific cases when we skip by early returning, which looks like "pass" on pytest, it would be clearer for the test to assume past_key_values exist and then have each model individually handle it within their test files. This way each model can either use an explicit skip reason with unittest.skip or implement their model equivalent test. This is opinion though - would be good to have thoughts from @ydshieh

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

V. nice and clean PR and tests ❤️

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -2315,6 +2395,17 @@ def contrastive_search(
...
 
         if return_dict_in_generate:
+            # Contrastive search works by forward looking at the next token, so we need to exclude it from
+            # `past_key_values` to be consistent with the other decoding methods
+            if model_kwargs.get("past_key_values") is not None:
+                past_key_values = []
+                for layer in model_kwargs["past_key_values"]:
+                    layer_past_key_values = []
+                    for item in layer:
+                        layer_past_key_values.append(item[..., :-1, :])
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:2405]: Guaranteed consistent item shape? No. Guaranteed that the second to last dimension is the sequence length? Yes :) 

Some models squash together or permutate the first two dimensions (by default, the cache is `(batch_size, num_heads, sequence_length, embed_size_per_head)`), Bloom and GPTBigCode aka StarCoder being the biggest offenders.

One of the tests I added is in the mixin and touches contrastive search, so the fact that this output can be correctly used for continuations is tested :D In fact, I only noticed that this was a problem precisely because of the test!

File: tests/generation/test_utils.py (hunk 1 above)
Comment (gante) [reply tests/generation/test_utils.py:1861]: I 100% agree!

Can I make it part of a follow-up PR? The same pattern is used in other places, and I am not sure how many models will break (there should be a few)

**Review by amyeroberts — APPROVED** [review]
Sorry - I meant to approve before!

File: tests/generation/test_utils.py (hunk 1 above)
Comment (amyeroberts) [reply tests/generation/test_utils.py:1861]: Yes, of course :D

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -1822,6 +1822,85 @@ def test_generate_from_inputs_embeds_decoder_only(self):
...
+
+            # If it doesn't support cache, pass the test
+            if not hasattr(config, "use_cache"):
+                return
+
+            # Let's make it always:
+            # 1. use cache (for obvious reasons)
+            # 2. generate to max length (which can be achieved by setting the eos token to an invalid value), which
+            #    would make the test flaky (e.g. EOS is generated on iteration 1 on both generations, but the
+            #    continuation would force it to generate beyond an EOS token)
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:1845]: Yes, correct -- it would make the test a false positive, failing on expected behavior.

**Comment by gante** [comment]
(merging and leaving the conversion to a skip as a TODO)

File: tests/generation/test_utils.py (hunk 1 above)
Comment (ydshieh) [reply tests/generation/test_utils.py:1861]: we can use `self.skipTest` (but generally not a good useage neither), especially here we are in a loop
```
for model_class
```
Also, better to use `continue` (compared to `return`) here.

Potentailly, using 

```
with self.subTest(....)
```
is a better approach I guess.

**Comment by nevakrien** [comment]
I dont see a version number when will this be out?

**Comment by amyeroberts** [comment]
@nevakrien 4.36v is now out :)
